[12.0][ADD] hr_attendance_split

patrickrwilson opened the pull request:
Related to PR #861 
This module was originally part of the attendance sheet but reviewers wished it to be in it's own standalone module, so this PR is for that purpose.

This is in alpha state as i'm sure adjustments will be needed as well as tests created in order for it to come out of alpha state.

patrickrwilson committed:
[ADD] hr_attendance_split

Initial Module

Updated Create and Readme

Update hr_attendance_split/models/hr_attendance.py

Co-authored-by: Maxime Chambreuil <mchambreuil@opensourceintegrators.com>

simplify if statement

Co-authored-by: Maxime Chambreuil <mchambreuil@opensourceintegrators.com>

Simplify if statement

Co-authored-by: Maxime Chambreuil <mchambreuil@opensourceintegrators.com>

Update hr_attendance_split/models/hr_attendance.py

Co-authored-by: Maxime Chambreuil <mchambreuil@opensourceintegrators.com>

simplify if statement

Co-authored-by: Maxime Chambreuil <mchambreuil@opensourceintegrators.com>

[FIX] Put split back into compute

Having the split calculations in the write and create caused continued problems so i've moved it into the compute function which seems to be working much better.

[FIX] references to attendance sheet method

[IMP] Move back to write/create

Moved functions back to write method. Overrides create to force user not to create attendance that goes overnight.

[FIX] settings wording

[FIX] comment in method

[FIX] split methods

This should fix the itssue with tz and splitting the attendances.

[FIX] Create Method

Fixed no longer used method within create method.

[FIX] Updated write method

Fixes issues with constraint checks.

patrickrwilson commented:
ping @jarroyomorales @max3903 @osi-scampbell

max3903 commented on hr_attendance_split/models/hr_attendance.py:
```diff
@@ -0,0 +1,102 @@
+# Copyright 2020 Pavlov Media
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+
+from odoo import api, fields, models, _
```
```suggestion
from odoo import api, fields, models
```

max3903 commented on hr_attendance_split/models/hr_attendance.py:
```diff
@@ -0,0 +1,102 @@
+                check_out = vals.get('check_out')
+            else:
+                check_out = rec.check_out
+            if 'check_in' in vals:
+                check_in = vals.get('check_in')
+            else:
+                check_in = rec.check_in
```
```suggestion
            check_in = vals.get('check_in', False) or rec.check_in
```

max3903 commented on hr_attendance_split/models/hr_attendance.py:
```diff
@@ -0,0 +1,102 @@
+        """ If the user clocks out after midnight, then it will split the
+        attendance at midnight of the employees timezone."""
+        for rec in self:
+            if 'check_out' in vals:
+                check_out = vals.get('check_out')
+            else:
+                check_out = rec.check_out
```
```suggestion
            check_out = vals.get('check_out', False) or rec.check_out
```

max3903 commented on hr_attendance_split/models/hr_attendance.py:
```diff
@@ -0,0 +1,102 @@
+    def create(self, vals):
+        """ If during creation, check-out crosses overnight, then make sure
+        the clock-out is at local midnight to prevent overnight attendances"""
+        if 'check_out' in vals and vals.get('check_out', False):
```
```suggestion
        if vals.get('check_out', False):
```

max3903 commented on hr_attendance_split/models/hr_attendance.py:
```diff
@@ -0,0 +1,102 @@
+
+    def _get_attendance_employee_tz(self, date=None):
+        """ Convert date according to timezone of user """
+        tz = False
+        if self.employee_id:
+            tz = self.employee_id.tz
+        if not date:
+            return False
```
```suggestion
        if not date:
            return False
        tz = False
        if self.employee_id:
            tz = self.employee_id.tz
```

jarroyomorales commented:
@patrickrwilson Hi! I was trying to test it on runbot but I got this error.
To reproduce, I created an attendance with a check in date, saved, and the added the checkout on the following day, expecting it to split the attendance in two.

![image](https://user-images.githubusercontent.com/47532076/93189742-cb5d1a00-f742-11ea-9cf0-b5a6540b318f.png)

patrickrwilson commented:
Thanks @jarroyomorales I can reproduce so i'll look into it.

patrickrwilson commented:
@jarroyomorales I think I fixed it. After fighting with the create and write methods I decided to change it and move it into the compute method, like I originally had it in the sheets module. It seems to work better for me.

jarroyomorales commented:
@patrickrwilson Hi! Im sorry but I am not being able to make it work. I have checked the 'split_attendances'  option but if i create an attendance with a check in date, save, and the add the checkout on the following day, it is not splitted. Im testing on runbot so maybe the error is due to another module incompatibility?

Another thing is that i dont find the `def _compute_duration(self)` function on the odoo core, so I dont know when it should be computed, is the name correct?

patrickrwilson commented:
@jarroyomorales I'm sorry, I don't know what I was thinking. For some reason I was referencing the hr_attendance_sheet method for duration but we wanted to have this standalone, so i'm not sure why I did that. 

The last commit works without the need for the attendance sheet module and should now be fully standalone. I'm not sure if the way I did it was the best way but it seems to work (I think).

jarroyomorales commented:
@patrickrwilson I think the best place to implement this functionality is in the `write` method of `hr_attendance` since writing and creating records on a compute function hardly ever is the best option. Also I think the field `need_split` is not necessary.
In hr_attendance's write function: `if vals.get('check_out', False): Check split`

jarroyomorales commented on hr_attendance_split/views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,27 @@
+                            <span class="fa fa-lg fa-building-o" title="Values set here are company-specific." aria-label="Values set here are company-specific." groups="base.group_multi_company" role="img"/>
+                            <div class="text-muted">
+                                Choose if attendances will be split if the start/end times cross over midnight.
+                                Example: User clocks in at 11pm, then clocks out 9am next day, two attendance would be created with the first clock-out at midnight and next clock-in at 12:01am.
```
This line is not correct is it? The module sets the check-out at 23:59, and the check-in at midnight (00:00).

patrickrwilson replied:
I missed this, you are correct. I corrected the wording.

patrickrwilson commented:
@jarroyomorales I originally put this in a write method however it wouldn't trigger if a user manually created an attendance that overlapped (Like a manager creating it for a missed day or something). It only worked during normal check-in then check-out. When I also tried to add it to the create method it caused errors and I think it was because there is also a create in the write method and as they looped it just didn't work right.

I'm fine with having it in the write method and just not in a create method if you are.

patrickrwilson commented:
@jarroyomorales Oh, I just remembered that the other problem with only having it in the write was that if the attendance was more than one day (Like check-in was 9/19 and check-out was 9/21), it wouldn't continue to split them for each day where as the compute does.

That was another reason why I abandoned the write/create methods. Ideas?

jarroyomorales commented:
Yes, thinking about it i think its correct if it doesnt trigger on manual creations with a check in and a check out. We could consider that case as forcing the dates.

And about the loop I think that if we make a version that splits the attendances properly this problem shouldn't appear. So tell me if you need anything!

jarroyomorales commented:
> @jarroyomorales Oh, I just remembered that the other problem with only having it in the write was that if the attendance was more than one day (Like check-in was 9/19 and check-out was 9/21), it wouldn't continue to split them for each day where as the compute does.
> 
> That was another reason why I abandoned the write/create methods. Ideas?

A dirty solution to this is when splitting an attendance and creating the 00:00 - XX:XX attendance, first create the attendance with the 00:00 check-in date and False checkout, and then write the XX:XX checkout date to trigger this again. I dont know if I explained myself properly.

jarroyomorales commented:
Also maybe it would be great to link the split attendances, to know which is the second part of an attendance.

patrickrwilson commented:
@jarroyomorales phew, I think I got it. This commit moves the function back to the write and your suggestion to write the check_out separately works like a charm. 

I added the create to just prevent the user from creating attendances that go overnight if they try to manually do it, which I thought was ok, since if they are manually adding attendances, then they probably should do it the manual way anyway. 

I also added the field to reference to the attendance it creates.

jarroyomorales commented on hr_attendance_split/models/hr_attendance.py:
```diff
@@ -0,0 +1,93 @@
+    @api.multi
+    def write(self, vals):
+        res = super().write(vals)
+        """ If the user clocks out after midnight, then it will split the
```
TravisCI is complaining about this line. I think this triple quote string should go on the first line of the function or as a comment with #

patrickrwilson replied:
I've moved the comment to the top.

jarroyomorales commented on hr_attendance_split/models/hr_attendance.py:
```diff
@@ -0,0 +1,93 @@
+        return attendance_tz_date_str
+
+    @api.model
+    def create(self, vals):
```
I am not sure on what to do here. I think we should allow a user to do this if he is manually creating it on purpose

patrickrwilson replied:
Here's my opinion, if the company wants split attendance then it shouldn't allow them to manual do it either. They would get the warning, then the have to manually create multiple attendances just like it automatically does when a user is clocking in/out on different days. I think if we let them enter any in there then it goes against the purpose of this module. 
If it is decided it's best to leave this out, it's easy to remove it.

jarroyomorales replied:
Okay, we can leave it like this for now then. Let's see if someone else has an opinion on this.

jarroyomorales commented:
@patrickrwilson H! I just tested it and it works as expected! Thank you very much!

The only case where it is not working properly is with attendances implying more than 2 days. I get the error `"Check Out" time cannot be earlier than "Check In" time. `. It may be because of timezones.

patrickrwilson commented:
> The only case where it is not working properly is with attendances implying more than 2 days. I get the error `"Check Out" time cannot be earlier than "Check In" time. `. It may be because of timezones.

@jarroyomorales I'm not able to reproduce this, it was able to do 5days worth of attendances. Can you provide the tz and start/end times you tried?
